app/models/audit/event/authn.rb

```diff
@@ -0,0 +1,53 @@
+require 'ostruct'
+
+class Audit::Event::Authn < OpenStruct
+  def emit_success
+    self.success = true
+    Audit.info message, 'authn', facility: 10, **structured_data
+  end
+
+  def emit_failure error_message
+    self.error_message = error_message
+    self.success = false
+    Audit.warn message, 'authn', facility: 10, **structured_data
+  end
+
+  def message
+    if success?
+      SUCCESS_TEMPLATE % [role_id, authenticator_name, service_id]
+    else
+      FAILURE_TEMPLATE % [role_id, authenticator_name, service_id, error_message]
+    end
+  end
+
+  SUCCESS_TEMPLATE = "%s successfully authenticated with authenticator %s service %s".freeze
+  FAILURE_TEMPLATE = "%s failed to authenticate with authenticator %s service %s: %s".freeze
+
+  def role_id
+    role.id
+  end
+
+  def service_id
+    service.try :id
```

Comment:
I'd prefer not to rely on Rails methods in our domain code.  I'd like to see a require active_support if you want to keep it (just to make that dep clear) or else `service.id rescue nil`

Reply:
I understand where you're coming from, but I like the expressiveness of some of ActiveSupport utilities and this code is unlikely to be moved out.